test/Test/Database/LSMTree/StateMachine.hs

```diff
@@ -465,97 +469,98 @@ createSystemTempDirectory prefix = do
+handleSessionDirDoesNotExistError = \case
+  ErrSessionDirDoesNotExist _dir -> Model.ErrSessionDirDoesNotExist
 
-checksumErrorHandler :: Monad m => Handler m (Maybe Model.Err)
-checksumErrorHandler = Handler $ pure . handler'
-  where
-    handler' :: ChecksumError -> Maybe Model.Err
-    handler' e = Just (Model.ErrSnapshotCorrupted (displayException e))
+handleSessionDirLockedError :: SessionDirLockedError -> Model.Err
+handleSessionDirLockedError = \case
+  ErrSessionDirLocked _dir -> Model.ErrSessionDirLocked
+
+handleSessionDirCorruptedError :: SessionDirCorruptedError -> Model.Err
+handleSessionDirCorruptedError = \case
+  ErrSessionDirCorrupted _dir -> Model.ErrSessionDirCorrupted
+
+handleSessionClosedError :: SessionClosedError -> Model.Err
+handleSessionClosedError = \case
+  ErrSessionClosed -> Model.ErrSessionClosed
+
+handleTableClosedError :: TableClosedError -> Model.Err
+handleTableClosedError = \case
+  ErrTableClosed -> Model.ErrTableClosed
+
+handleTableCorruptedError :: TableCorruptedError -> Model.Err
+handleTableCorruptedError = \case
+  ErrLookupByteCountDiscrepancy _ _ -> Model.ErrTableCorrupted
 
--- | When combined with other handlers, 'catchAllErrorHandler' has to go last
--- because it matches on 'SomeException', and no other handlers are run after
--- that. See the use of 'catches' in 'catchErr'.
-catchAllErrorHandler :: Monad m => Handler m (Maybe Model.Err)
-catchAllErrorHandler = Handler $ \(e :: SomeException) ->
-    pure $ Just (Model.ErrOther (displayException e))
+handleTableNotCompatibleError :: TableNotCompatibleError -> Model.Err
+handleTableNotCompatibleError = \case
+  ErrTableTypeMismatch _ _ -> Model.ErrTableTypeMismatch
+  ErrTableSessionMismatch _ _ -> Model.ErrTableSessionMismatch
+
+handleSnapshotExistsError :: SnapshotExistsError -> Model.Err
+handleSnapshotExistsError = \case
+  ErrSnapshotExists name -> Model.ErrSnapshotExists name
+
+handleSnapshotDoesNotExistError :: SnapshotDoesNotExistError -> Model.Err
+handleSnapshotDoesNotExistError = \case
+  ErrSnapshotDoesNotExist name -> Model.ErrSnapshotDoesNotExist name
+
+handleSnapshotCorruptedError :: SnapshotCorruptedError -> Model.Err
+handleSnapshotCorruptedError = \case
+  ErrSnapshotCorrupted name _ -> Model.ErrSnapshotCorrupted name
+
+handleSnapshotNotCompatibleError :: SnapshotNotCompatibleError -> Model.Err
+handleSnapshotNotCompatibleError = \case
+  ErrSnapshotWrongTableType name _ _ -> Model.ErrSnapshotWrongTableType name
+  ErrSnapshotWrongLabel name _ _ -> Model.ErrSnapshotWrongLabel name
+
+handleBlobRefInvalidError :: BlobRefInvalidError -> Model.Err
+handleBlobRefInvalidError = \case
+  ErrBlobRefInvalid _ -> Model.ErrBlobRefInvalid
+
+handleCursorClosedError :: CursorClosedError -> Model.Err
+handleCursorClosedError = \case
+  ErrCursorClosed -> Model.ErrCursorClosed
+
+handleFsError :: FsError -> Model.Err
+handleFsError = Model.ErrDiskFault . displayException
```

Comment:
`ErrDiskFault` was originally representing a supserset of errors. Maybe we can map `FsError` to `ErrFsError`, and then reserve `ErrDiskFault` for the catch-all "something went wrong" error that the model throws

Reply:
So map FsError to ErrFsError and then match ErrFsError and the model's ErrDiakFault?